### PR TITLE
Support for streamed responses

### LINKED ISSUE #75: No akka_http_responses_size_bytes_bucket for streamed endpoints
Hello, with akka-http-metrics-prometheus v1.3.0

I noticed that when an endpoint responds with streamed data, for ex:
```scala
pathLabeled("testStream", "testStream") {
  get {
     complete(Source(List("a","b","c")))
  }
} 
```

There is no `akka_http_responses_size_bytes_bucket` metric for label `testStream`, however for non-streamed endpoints it is computed correctly.

Here is an example of `/metrics` output after 1 call to `/testStream` and 1 call to `/metrics` (only method and path dimensions are enabled):

```
# HELP akka_http_responses_duration_seconds HTTP response duration
# TYPE akka_http_responses_duration_seconds histogram
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.005",} 0.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.01",} 0.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.025",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.05",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.075",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.1",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.25",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="0.75",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="1.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="2.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="5.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="7.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="10.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/testStream",le="+Inf",} 1.0
akka_http_responses_duration_seconds_count{method="GET",path="/testStream",} 1.0
akka_http_responses_duration_seconds_sum{method="GET",path="/testStream",} 0.024
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.005",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.01",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.025",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.05",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.075",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.1",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.25",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="0.75",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="1.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="2.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="5.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="7.5",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="10.0",} 1.0
akka_http_responses_duration_seconds_bucket{method="GET",path="/metrics",le="+Inf",} 1.0
akka_http_responses_duration_seconds_count{method="GET",path="/metrics",} 1.0
akka_http_responses_duration_seconds_sum{method="GET",path="/metrics",} 0.0
# HELP akka_http_requests_active Active HTTP requests
# TYPE akka_http_requests_active gauge
akka_http_requests_active 0.0
# HELP akka_http_requests_size_bytes HTTP request size
# TYPE akka_http_requests_size_bytes histogram
akka_http_requests_size_bytes_bucket{le="0.0",} 2.0
akka_http_requests_size_bytes_bucket{le="100.0",} 2.0
akka_http_requests_size_bytes_bucket{le="200.0",} 2.0
akka_http_requests_size_bytes_bucket{le="300.0",} 2.0
akka_http_requests_size_bytes_bucket{le="400.0",} 2.0
akka_http_requests_size_bytes_bucket{le="500.0",} 2.0
akka_http_requests_size_bytes_bucket{le="600.0",} 2.0
akka_http_requests_size_bytes_bucket{le="700.0",} 2.0
akka_http_requests_size_bytes_bucket{le="800.0",} 2.0
akka_http_requests_size_bytes_bucket{le="900.0",} 2.0
akka_http_requests_size_bytes_bucket{le="1000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="2000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="3000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="4000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="5000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="6000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="7000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="8000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="9000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="10000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="20000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="30000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="40000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="50000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="60000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="70000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="80000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="90000.0",} 2.0
akka_http_requests_size_bytes_bucket{le="+Inf",} 2.0
akka_http_requests_size_bytes_count 2.0
akka_http_requests_size_bytes_sum 0.0
# HELP akka_http_requests_total Total HTTP requests
# TYPE akka_http_requests_total counter
akka_http_requests_total 2.0
# HELP akka_http_responses_size_bytes HTTP response size
# TYPE akka_http_responses_size_bytes histogram
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="0.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="100.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="200.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="300.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="400.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="500.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="600.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="700.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="800.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="900.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="1000.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="2000.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="3000.0",} 0.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="4000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="5000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="6000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="7000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="8000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="9000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="10000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="20000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="30000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="40000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="50000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="60000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="70000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="80000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="90000.0",} 1.0
akka_http_responses_size_bytes_bucket{method="GET",path="/metrics",le="+Inf",} 1.0
akka_http_responses_size_bytes_count{method="GET",path="/metrics",} 1.0
akka_http_responses_size_bytes_sum{method="GET",path="/metrics",} 3827.0
# HELP akka_http_responses_total HTTP responses
# TYPE akka_http_responses_total counter
akka_http_responses_total{method="GET",path="/testStream",} 1.0
akka_http_responses_total{method="GET",path="/metrics",} 1.0
```

### PR DESCRIPTION
**Pull request by Fraer**
Add support for streamed responses to `responsesSize` histogram, if response entity `contentLength` is unknown apply transforming flow that accumulates the sum of every chunk size then once the stream is complete (see `akka.stream.scaladsl.FlowOpsMat.watchTermination`) update the responsesSize histogram.

Fixes #75

### CODE REVIEW

**Review by RustedBones — COMMENTED** [review]
I think we should consume the stream instead of transforming it. An error in the metrics lib won't have an impact on the server.
And since data is immutable there should not be any copy of the byte stream.

This should also be applied for request size when data is given as stream.

File: core/src/main/scala/fr/davit/akka/http/metrics/core/HttpMetricsHandler.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ import scala.concurrent.{ExecutionContext, Future}
 
 trait HttpMetricsHandler {
 
-  def onRequest(request: HttpRequest, response: Future[HttpResponse])(implicit executionContext: ExecutionContext): Unit
+  def onRequest(request: HttpRequest, response: Future[HttpResponse])(implicit executionContext: ExecutionContext): Future[HttpResponse]
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply core/src/main/scala/fr/davit/akka/http/metrics/core/HttpMetricsHandler.scala]: It actually will require a `Materializer`

**Comment by RustedBones** [comment]
@Fraer WDYT of the changes ?

**Comment by Fraer** [comment]
Just tested your changes and unfortunately my whole app stopped working :(

Got many errors like :
```
server 19:49:25.480 [default-akka.actor.default-dispatcher-16] ERROR akka.actor.RepointableActorRef -  Error in stage [akka.http.impl.util.StreamUtils$$anon$1@acafd58]: Deflater has been closed
server java.lang.NullPointerException: Deflater has been closed
server 	at java.base/java.util.zip.Deflater.ensureOpen(Deflater.java:898)
server 	at java.base/java.util.zip.Deflater.deflate(Deflater.java:566)
server 	at java.base/java.util.zip.Deflater.deflate(Deflater.java:484)
...
```

I guess consuming the stream to compute the size makes the stream unavailable for second consumption by the remote caller. I'm not an expert in akka-streams but i guess a stream can be consumed only once 😢

**Comment by Fraer** [comment]
maybe consuming data to get the size could be moved into a separate flow with `alsoTo`

**Comment by RustedBones** [comment]
@Fraer thanks a lot for checking. Indeed we can't run the stream twice.

I'm thinking to change the whole API, leveraging stream attributes to be able to support this in a more transparent way. Hopefully this will also simplify the code.

**Comment by RustedBones** [comment]
This looked ready for the 1.4.0 release. I tried a minimal test with streams and it worked on my side.
Can you have can you have another look @Fraer ? If it works for you, I'll prepare the release ASAP.

**Comment by Fraer** [comment]
Just tested with the 1.4.0-SNAPSHOT and on first call to my app i got: 
```
server java.lang.IllegalArgumentException: Incorrect number of labels.
server 	at io.prometheus.client.SimpleCollector.labels(SimpleCollector.java:64)
server 	at fr.davit.akka.http.metrics.prometheus.PrometheusGauge.inc(PrometheusMetrics.scala:31)
server 	at fr.davit.akka.http.metrics.core.HttpMetricsRegistry.onRequest(HttpMetricsRegistry.scala:152)
server 	at fr.davit.akka.http.metrics.core.HttpMetrics$.$anonfun$meterFunction$3(HttpMetrics.scala:99)
server 	at scala.Function1.$anonfun$andThen$1(Function1.scala:57)
server 	at scala.Function1.$anonfun$andThen$1(Function1.scala:57)
server 	at scala.Function1.$anonfun$andThen$1(Function1.scala:57)
server 	at fr.davit.akka.http.metrics.core.HttpMetrics$.$anonfun$meterFunction$1(HttpMetrics.scala:107)
```

Everything works with 1.3.0

**Comment by RustedBones** [comment]
Thanks again @Fraer.

I know where the issue comes from. I made a change to dimensions without updating Prometheus registry implementation. It required to know the dimension keys before accepting data.

**Review by RustedBones — APPROVED** [review]

**Pull request merged**